lib/web_ui/dev/firefox_installer.dart

```diff
@@ -0,0 +1,136 @@
+    );
+    final io.Directory versionDir = io.Directory(
+      path.join(firefoxInstallationDir.path, version),
+    );
+    return FirefoxInstaller._(
+      version: version,
+      firefoxInstallationDir: firefoxInstallationDir,
+      versionDir: versionDir,
+    );
+  }
+
+  static Future<FirefoxInstaller> latest() async {
+    final String latestVersion = await fetchLatestFirefoxVersion();
+    return FirefoxInstaller(version: latestVersion);
+  }
+
+  FirefoxInstaller._({
+    @required this.version,
+    @required this.firefoxInstallationDir,
+    @required this.versionDir,
+  });
+
+  /// Firefox version managed by this installer.
+  final String version;
+
+  /// HTTP client used to download Firefox.
+  final Client client = Client();
+
+  /// Root directory that contains Chrome versions.
+  final io.Directory firefoxInstallationDir;
+
+  /// Installation directory for Firefox of the requested [version].
+  final io.Directory versionDir;
+
+  bool get isInstalled {
+    return versionDir.existsSync();
+  }
+
+  BrowserInstallation getInstallation() {
+    if (!isInstalled) {
+      return null;
+    }
+
+    return BrowserInstallation(
+      version: version,
+      executable: PlatformBinding.instance.getFirefoxExecutablePath(versionDir),
+    );
+  }
+
+  /// Install the browser by downloading from the web.
+  Future<void> install() async {
+    final io.File downloadedFile = await _download();
+    await _uncompress(downloadedFile);
+    downloadedFile.deleteSync();
+  }
+
+  /// Downloads the browser version from web.
+  /// See [version].
+  Future<io.File> _download() async {
+    if (versionDir.existsSync()) {
+      versionDir.deleteSync(recursive: true);
+    }
+
+    versionDir.createSync(recursive: true);
+    final String url = PlatformBinding.instance.getFirefoxDownloadUrl(version);
+    final StreamedResponse download = await client.send(Request(
+      'GET',
+      Uri.parse(url),
+    ));
+
+    final io.File downloadedFile =
+        io.File(path.join(versionDir.path, 'firefox.zip'));
+    await download.stream.pipe(downloadedFile.openWrite());
+
+    return downloadedFile;
+  }
+
+  /// Uncompress the downloaded browser files.
+  /// See [version].
+  Future<void> _uncompress(io.File downloadedFile) async {
+    /// TODO(nturgut): Implement Install.
+  }
+
+  void close() {
+    client.close();
+  }
+}
+
+/// Fetches the latest available Chrome build version.
+Future<String> fetchLatestFirefoxVersion() async {
+  final RegExp forFirefoxVersion = RegExp("firefox-[0-9.]\+[0-9]");
+  final io.HttpClientRequest request = await io.HttpClient()
+      .getUrl(Uri.parse(PlatformBinding.instance.getFirefoxLatestVersionUrl()));
+  request.followRedirects = false;
+  // We will parse the HttpHeaders to find the redirect location.
+  final io.HttpClientResponse response = await request.close();
+
+  String version;
+
+  response.headers.forEach((String s, List<String> list) {
```

Comment:
I think you can do `response.headers.value('location')`, no? See: https://api.flutter.dev/flutter/dart-io/HttpHeaders/value.html

Reply:
Thanks! This is a lot nicer.

Reply:
Done.